src/xmipp/libraries/reconstruction/angular_sph_alignment.cpp

```diff
@@ -474,122 +481,157 @@ void ProgAngularSphAlignment::writeImageParameters(const FileName &fnImg) {
 {
-	size_t groups = 4;
-	size_t olditems = (4*prevL)/groups;
-	size_t newitems = (4*L)/groups;
-	for (int g=0;g<groups;g++)
-	{
-		for (int i=0;i<olditems;i++)
-			{
-			    newvect(g*newitems+i) = oldvect(g*olditems+i);
-			}
-	}
-	for (int i=0;i<5;i++)
-	{
-		newvect(groups*newitems+i) = oldvect(groups*olditems+i);
-	}
+    int idx = 0;
+	vL1.initZeros(vecSize);
+	vN.initZeros(vecSize);
+	vL2.initZeros(vecSize);
+	vM.initZeros(vecSize);
+    for (int h=0; h<=l2; h++) {
+        int totalSPH = 2*h+1;
+        int aux = std::floor(totalSPH/2);
+        for (int l=h; l<=l1; l+=2) {
+            for (int m=0; m<totalSPH; m++) {
+                VEC_ELEM(vL1,idx) = l;
+                VEC_ELEM(vN,idx) = h;
+                VEC_ELEM(vL2,idx) = h;
+                VEC_ELEM(vM,idx) = m-aux;
+                idx++;
+            }
+        }
+    }
+}
+
+void ProgAngularSphAlignment::updateCTFImage(double defocusU, double defocusV, double angle)
+{
+	ctf.K=1; // get pure CTF with no envelope
+	currentDefocusU=ctf.DeltafU=defocusU;
+	currentDefocusV=ctf.DeltafV=defocusV;
+	currentAngle=ctf.azimuthal_angle=angle;
+	ctf.produceSideInfo();
 }
 
-void ProgAngularSphAlignment::deformVol(MultidimArray<double> &mVD, const MultidimArray<double> &mV, double &def)
+void ProgAngularSphAlignment::deformVol(MultidimArray<double> &mP, const MultidimArray<double> &mV, double &def,
+                                        double rot, double tilt, double psi)
 {
-	int l,n,m;
-	size_t idxY0=(VEC_XSIZE(clnm)-5)/4;
+    int l1,n,l2,m;
+	l1 = 0;
+	n = 0;
+	l2 = 0;
+	m = 0;
+	size_t idxY0=(VEC_XSIZE(clnm)-8)/3;
 	double Ncount=0.0;
-	double totalVal=0.0;
-	double voxModg=0.0;
+    double modg=0.0;
 	double diff2=0.0;
 
 	def=0.0;
 	size_t idxZ0=2*idxY0;
-	size_t idxR=3*idxY0;
 	sumVd=0.0;
+	double RmaxF=RmaxDef;
+	double RmaxF2=RmaxF*RmaxF;
+	double iRmaxF=1.0/RmaxF;
+    // Rotation Matrix
+    Matrix2D<double> R;
+    R.initIdentity(3);
+    Euler_angles2matrix(rot, tilt, psi, R, false);
+    R = R.inv();
+    Matrix1D<double> pos;
+    pos.initZeros(3);
+
+	// TODO: Poner primero i y j en el loop, acumular suma y guardar al final
 	for (int k=STARTINGZ(mV); k<=FINISHINGZ(mV); k++)
 	{
 		for (int i=STARTINGY(mV); i<=FINISHINGY(mV); i++)
 		{
 			for (int j=STARTINGX(mV); j<=FINISHINGX(mV); j++)
 			{
-
+                ZZ(pos) = k; YY(pos) = i; XX(pos) = j;
+                pos = R * pos;
 				double gx=0.0, gy=0.0, gz=0.0;
-				for (size_t idx=0; idx<idxY0; idx++)
-				{
-					double RmaxF=VEC_ELEM(clnm,idx+idxR);
-					double RmaxF2=RmaxF*RmaxF;
-					double iRmaxF=1.0/RmaxF;
-					double k2=k*k;
-					double kr=k*iRmaxF;
-					double k2i2=k2+i*i;
-					double ir=i*iRmaxF;
-					double r2=k2i2+j*j;
-					double jr=j*iRmaxF;
-					double zsph=0.0;
-					double rr=sqrt(r2)*iRmaxF;
-					if (r2<RmaxF2)
-					{
-						spherical_index2lnm(idx+1,l,n,m);
-						zsph=ZernikeSphericalHarmonics(l,n,m,jr,ir,kr,rr);
+				// TODO: Sacar al bucle de z
+				double k2=ZZ(pos)*ZZ(pos);
```

Comment:
Since we need to rotate the k,i,j positions before computing k2 and so on, it is safer to use the pos vector. Since these positions are float in general, probably the copy of the modified vector back to variables k,i,j might introduce errors in the loop